### PR TITLE
Aquifer fetkovich

### PR DESCRIPTION
**Pull request by WesselDeZeeuw**
Hi guys,

Here is the start of my Fetkovich Aquifer implementation. This is still a Work In Progress! This pull request for now is so that @nairr and @kel85uk can easily see my code and comment on it! 

Thanks guys, will keep you updated

### CODE REVIEW

**Comment by nairr** [comment]
@hnil - we are currently testing this Fetkovich implementation on the 2D_OILWATER model (https://github.com/OPM/opm-tests/tree/master/aquifer-oilwater) and SPE1CASE2. The aquifer solution is currently not a match with ECL but the idea is to have a working implementation by the end of the year or early next year

**Comment by hnil** [comment]
We have implemented a version in MRST. If you could share the test case with ecl output it would be very valuable for us. We could give you access to the MRST implementation which probably will be usefull for debugging..

**Comment by WesselDeZeeuw** [comment]
Dear @hnil 
Here you find the ECL results together with the .data file we have used to produce this, Hope this helps!
Unfortunately, GitHub doesn't allow these file formats, so everything is in a .zip file!

Cheers! 
Wessel

[aqufetp_eclipse.zip](https://github.com/OPM/opm-simulators/files/2594246/aqufetp_eclipse.zip)

**Comment by hnil** [comment]
Thanks. Maybe my comments on the other aquifer model also is of interest. We are trying to get a quite complicated reservoir model with aquifers to run.

**Comment by WesselDeZeeuw** [comment]
The Aquifer Fetkovich Implementation is ready just before Christmas. We have added a test case in opm-tests OPM/opm-tests#83

This pull-requests makes the necessary changes for the Fetkovich Aquifer and the Test Case
Please make a review of the code. 

Have the best holidays.
Wessel

**Comment by GitPaean** [comment]
jenkins build this opm-tests=83 please

**Comment by GitPaean** [comment]
@WesselDeZeeuw , thanks for the contribution.  It will take some time to discuss this PR, so it is mostly likely will get in after New Year. 

The firs question here is, you attached an example https://github.com/OPM/opm-simulators/files/2594246/aqufetp_eclipse.zip above.  To me, this PR does not produce the included reference results. Do you have any clue about the cause?

**Comment by GitPaean** [comment]
I just notice that the above example only contains one report step for one day, which makes it difficult to judge whether the result is good or bad. 

Can you confirm this example by running it for longer time?

**Comment by WesselDeZeeuw** [comment]
> I just notice that the above example only contains one report step for one day, which makes it difficult to judge whether the result is good or bad.
> 
> Can you confirm this example by running it for longer time?

If you take al ook at PR OPM/opm-tests#83 there you will find the correct model. The files included in the .zip file above is a test case which we used to consider previously, but we now have updated. This example is also run for 60 timesteps!

Take your time to review this contribution. I will be on vacation from today untill jan 3rd, so I will reply to this, from that time.

Cheers,
Wessel

**Comment by GitPaean** [comment]
> If you take al ook at PR OPM/opm-tests#83 there you will find the correct model. The files included in the .zip file above is a test case which we used to consider previously, but we now have updated. This example is also run for 60 timesteps!

I have tested the example in  OPM/opm-tests#83, it works fine. I just want to confirm that the above example is not relevant anymore. Just out of curiosity, what is the problem of the above example?

> Take your time to review this contribution. I will be on vacation from today untill jan 3rd, so I will reply to this, from that time.

Have a nice vacation. We have a similar timeline, so we will discuss after the holiday.

File: tests/update_reference_data.sh [hunk 1]
```diff
@@ -38,6 +38,7 @@ tests[spe1oilgas]="flow spe1 SPE1CASE2_OILGAS spe1_oilgas"
 tests[spe1nowells]="flow spe1 SPE1CASE2_NOWELLS spe1_nowells"
 tests[spe1thermal]="flow spe1 SPE1CASE2_THERMAL spe1_thermal"
 tests[ctaquifer_2d_oilwater]="flow aquifer-oilwater 2D_OW_CTAQUIFER ctaquifer_2d_oilwater"
+tests[fetkovich_2d]= "flow aquifer-fetkovich 22D_FETKOVICHAQUIFER fetkovich_2d"
```
Comment (atgeirr) [inline tests/update_reference_data.sh]: The deck name should probably only have one "2"?

**Comment by GitPaean** [comment]
@WesselDeZeeuw , there is some merging conflict now. Please fix them.

**Comment by WesselDeZeeuw** [comment]
> @WesselDeZeeuw , there is some merging conflict now. Please fix them.

The merging conflicts are fixed. I also adapted your changes for grid_index in CarterTracy to the Fetkovich case. The results remain the same

**Comment by GitPaean** [comment]
> @GitPaean The merging conflicts are fixed. I also adapted your changes for grid_index in CarterTracy to the Fetkovich case. The results remain the same

It is great.  Could you please help to confirm with the changes related to grid_index, the results look okay for the attached case?

[2D_FETKOVICHAQUIFER_MOD.zip](https://github.com/OPM/opm-simulators/files/2723928/2D_FETKOVICHAQUIFER_MOD.zip)

The changes are that I added ACTNUM
```
ACTNUM
4*1
1*1
5*0
5*1
5*1
20*1
10*0
10*1
10*1
10*1
20*1
/
```
and also changed the 
```
AQUANCON
-- Aq#     I1 I2  J1 J2  K1 K2 FACE
    1      18 18   1  1  1  5  J- 1.0 1.0 NO/
```

Without ACTNUM, the grid_index related change will not impact anything.

File: opm/autodiff/AquiferFetkovich.hpp [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+            Opm::FaceDir::DirEnum faceDirection;
+
+            // denom_face_areas is the sum of the areas connected to an aquifer
+            Scalar denom_face_areas = 0.;
+            cellToConnectionIdx_.resize(ebos_simulator_.gridView().size(/*codim=*/0), -1);
+            for (size_t idx = 0; idx < cell_idx_.size(); ++idx)
+            {
+              cellToConnectionIdx_[cell_idx_[idx]] = idx;
+              const int cell_index = cartesian_to_compressed_.at(cell_idx_[idx]);
+              auto cellFacesRange = cell2Faces[cell_index];
```
Comment (GitPaean) [inline opm/autodiff/AquiferFetkovich.hpp]: The above code should be something as follows, 
```
                    const int cell_index = cartesian_to_compressed_.at(cell_idx_[idx]);
                    cellToConnectionIdx_[cell_index] = idx;
                    const auto cellFacesRange = cell2Faces[cell_index];
```

**Comment by WesselDeZeeuw** [comment]
> It is great. Could you please help to confirm with the changes related to grid_index, the results look okay for the attached case?
> 
> [2D_FETKOVICHAQUIFER_MOD.zip](https://github.com/OPM/opm-simulators/files/2723928/2D_FETKOVICHAQUIFER_MOD.zip)

M. I unfortunately get some different results (see the ZIP file for my Eclipse results and FLOW results). However, I also get these results whenever I revert back to the code without the adjustments. 

[2D_FETKOVICHAQUIFER_MOD.zip](https://github.com/OPM/opm-simulators/files/2724153/2D_FETKOVICHAQUIFER_MOD.zip)
 Any Ideas?

File: opm/autodiff/AquiferFetkovich.hpp (hunk 2 above)
Comment (WesselDeZeeuw) [reply opm/autodiff/AquiferFetkovich.hpp]: Changed!

**Comment by GitPaean** [comment]
I tested with the current PR, the results look pretty okay. 
[ftaq-0104.pdf](https://github.com/OPM/opm-simulators/files/2726386/ftaq-0104.pdf)

> However, I also get these results whenever I revert back to the code without the adjustments.

I am not sure I am following this.

File: opm/autodiff/AquiferFetkovich.hpp [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+        // This function implements Eq 5.14 of the EclipseTechnicalDescription
+        inline void calculateInflowRate(int idx, const Simulator& simulator)
+        {
+          Tc_ = ( aqufetp_data_.C_t * aqufetp_data_.V0 ) / aqufetp_data_.J ;
+          Scalar td_Tc_ = simulator.timeStepSize() / Tc_ ;
+          Scalar exp_ = (1 - exp(-td_Tc_)) / td_Tc_;
+          Qai_.at(idx) = alphai_.at(idx) * aqufetp_data_.J * dpai(idx) * exp_;
+        }
+
+        template<class faceCellType, class ugridType>
```
Comment (GitPaean) [inline opm/autodiff/AquiferFetkovich.hpp]: This function looks like the same with the same function from AquiferCarterTracy.hpp, and it does not use any member or functions from the class. Maybe you can put this type of function to header file like `AquiferHelpers.hpp`. 

You can refer to `WellHelpers.hpp`.

File: opm/autodiff/AquiferFetkovich.hpp [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+              faceArea = (cellNeighbour0 < 0)? faceArea : 0.;
+            }
+            else if (cellNeighbour0 == 0){
+              faceArea = (cellNeighbour1 < 0)? faceArea : 0.;
+            }
+            return faceArea;
+          }
+
+          // This function is used to initialize and calculate the alpha_i for each grid connection to the aquifer
+          inline void initializeConnections(const Aquancon::AquanconOutput& connection)
```
Comment (GitPaean) [inline opm/autodiff/AquiferFetkovich.hpp]: How different is this function from the version in `AquiferCarterTracy`. If they are the same, maybe with some refactoring, you can also put this function to the Helper file to reduce some duplication of code. Alternatively, you can put it in a base class for `AquiferFetkovich` and `AquiferCarterTracy`.

**Comment by GitPaean** [comment]
Can you please do `dos2unix` to file `BlackoilAquiferModel_impl.hpp`, to avoid change of the line breaks back ? It just causes problem for the `git diff`.

**Comment by GitPaean** [comment]
> Or set core.autocrlf=true in your (global) .gitconfig file.

But the problem stays. The git history will contain many not very `real` modifications.  I will follow whatever we agree in term of formatting.

**Comment by WesselDeZeeuw** [comment]
> The classes `AquiferCarterTracy` and `AquiferFetkovich` have a lot of duplicated functions. Maybe the proper way to do this is to create a base class for `AquiferCarterTracy` and `AquiferFetkovich` to reduce the code duplication.
> 
> The base class will have the most of the implementation, and `AquiferCarterTracy` and `AquiferFetkovich` as derived classes will have much less content, it looks like very few functions.

Yes, I'll do some more coding into this. Will let you know when it's ready! 


> Or set `core.autocrlf=true` in your (global) `.gitconfig` file.

Also did this, and will try to change this file dos2unix (not sure how, but will manage)

**Comment by WesselDeZeeuw** [comment]
Hi all,
I was wondering if it was possible to merge this PR first and to re-factor the code later as in PR #1695 . 
In my view this PR is for the functionality of the Aquifer, and the other PR is for restructuring the code. Let me know,

Thanks. Wessel

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=86 please

**Comment by GitPaean** [comment]
jenkins build this update_data please

**Comment by WesselDeZeeuw** [comment]
> I do not understand why there is no data updated for this PR in [OPM/opm-tests#86](https://github.com/OPM/opm-tests/pull/86) .

Is there anything I can do? :-)

**Comment by GitPaean** [comment]
> Is there anything I can do? :-)

I did not spot the cause in this PR. The worst case, we just create the reference result from the opm-tests side manually.

**Comment by GitPaean** [comment]
Creating reference result manually is a task can be done with a few minutes. I do not consider it a bad thing. However, it is not the preferred way we are dealing with it now. 

If needed, for sure we can do that from my side.

**Comment by WesselDeZeeuw** [comment]
> Creating reference result manually is a task can be done with a few minutes. I do not consider it a bad thing. However, it is not the preferred way we are dealing with it now.
> 
> If needed, for sure we can do that from my side.

I'm fine with whatever option. Preferably before the end of next week because of the OPM summit coming up ;-)

**Comment by GitPaean** [comment]
jenkins build this opm-tests=87 please

**Comment by GitPaean** [comment]
jenkins build this opm-tests=87 please

**Comment by GitPaean** [comment]
Thanks for the contribution. A re-factoring PR coming later to reduce the code duplication will be okay. 

Merging.

**Comment by GitPaean** [comment]
did not catch the following introduced warning. Should be fixed in the coming PR. 
```
/home/kaib/OPM-master-test/debug/opm-simulators/opm/autodiff/AquiferFetkovich.hpp:147:38: warning: ‘Opm::AquiferFetkovich<Ewoms::Properties::TTag::EclFlowSolventProblem>::aqufetp_data_’ will be initialized after [-Wreorder]
         const Aquifetp::AQUFETP_data aqufetp_data_;
                                      ^
/home/kaib/OPM-master-test/debug/opm-simulators/opm/autodiff/AquiferFetkovich.hpp:131:45: warning:   ‘const std::unordered_map<int, int>& Opm::AquiferFetkovich<Ewoms::Properties::TTag::EclFlowSolventProblem>::cartesian_to_compressed_’ [-Wreorder]
         const std::unordered_map<int, int>& cartesian_to_compressed_;
                                             ^
/home/kaib/OPM-master-test/debug/opm-simulators/opm/autodiff/AquiferFetkovich.hpp:66:5: warning:   when initialized here [-Wreorder]
     AquiferFetkovich( const Aquifetp::AQUFETP_data& aqufetp_data,
```

**Comment by WesselDeZeeuw** [comment]
yes, will focus on this too. Do you now what causes this issue? Because then I know where to focus on.
Cheers!

**Comment by bska** [comment]
> Do you now what causes this issue?

The C++ language says that data members of a class are initialised in the order of their declaration within the class definition.  Class `AquiferFetkovich` declares `compressed_to_cartesian_` ahead of `aqufetp_data_`.  Therefore we should (generally) initialise `compressed_to_cartesian_` ahead of `aqufetp_data_` within the constructor's initialiser list.

File: tests/update_reference_data.sh [hunk 5]
```diff
@@ -38,6 +38,7 @@ tests[spe1oilgas]="flow spe1 SPE1CASE2_OILGAS spe1_oilgas"
 tests[spe1nowells]="flow spe1 SPE1CASE2_NOWELLS spe1_nowells"
 tests[spe1thermal]="flow spe1 SPE1CASE2_THERMAL spe1_thermal"
 tests[ctaquifer_2d_oilwater]="flow aquifer-oilwater 2D_OW_CTAQUIFER ctaquifer_2d_oilwater"
+tests[fetkovich_2d]= "flow aquifer-fetkovich 2D_FETKOVICHAQUIFER fetkovich_2d"
```
Comment (GitPaean) [inline tests/update_reference_data.sh]: I noticed the grammar here is not that consistent.  For each entry in `tests`, should we have four elements or three. Above this line, there are four elements. below this line, there are three. Looking at usage at the bottom, it should contain 4 elements for each entry? @akva2

**Pull request merged**